src/WinRT.Runtime2/InteropServices/ExceptionHelpers.cs

```diff
@@ -0,0 +1,264 @@
+            try
+            {
+                if (WellKnownErrorCodes.Succeeded(ILanguageExceptionErrorInfo2Vftbl.GetPropagationContextHeadUnsafe((void*)languageErrorInfo2Ptr, &currentLanguageExceptionErrorInfo2Ptr)))
+                {
+                    while (currentLanguageExceptionErrorInfo2Ptr != null)
+                    {
+                        Exception? propagatedException = GetLanguageExceptionInternal(currentLanguageExceptionErrorInfo2Ptr, hr);
+                        if (propagatedException is not null)
+                        {
+                            return propagatedException;
+                        }
+
+                        void* previousLanguageExceptionErrorInfo2Ptr = currentLanguageExceptionErrorInfo2Ptr;
+                        try
+                        {
+                            _ = ILanguageExceptionErrorInfo2Vftbl.GetPreviousLanguageExceptionErrorInfoUnsafe(currentLanguageExceptionErrorInfo2Ptr, &currentLanguageExceptionErrorInfo2Ptr);
+                        }
+                        finally
+                        {
+                            _ = Marshal.Release((nint)previousLanguageExceptionErrorInfo2Ptr);
+                        }
+                    }
+                }
+            }
+            finally
+            {
+                MarshalExtensions.ReleaseIfNotNull(currentLanguageExceptionErrorInfo2Ptr);
+                _ = Marshal.Release(languageErrorInfo2Ptr);
+            }
+        }
+
+        return null;
+    }
+
+    private static unsafe Exception? GetLanguageExceptionInternal(void* languageErrorInfoPtr, int hr)
+    {
+        void* languageExceptionPtr = null;
+        try
+        {
+            if (WellKnownErrorCodes.Succeeded(ILanguageExceptionErrorInfoVftbl.GetLanguageExceptionUnsafe(languageErrorInfoPtr, &languageExceptionPtr)) && languageExceptionPtr != null)
+            {
+                if (WindowsRuntimeMarshal.IsReferenceToManagedObject(languageExceptionPtr))
+                {
+                    Exception ex = ComInterfaceDispatch.GetInstance<Exception>((ComInterfaceDispatch*)languageExceptionPtr);
+                    if (GetHRForException(ex) == hr)
+                    {
+                        return ex;
+                    }
+                }
+            }
+        }
+        finally
+        {
+            _ = Marshal.Release((nint)languageExceptionPtr);
+        }
+
+        return null;
+    }
+
+    public static unsafe int GetHRForException(Exception ex)
+    {
+        int hr = ex.HResult;
+        try
+        {
+            if (ex.TryGetRestrictedLanguageErrorInfo(out WindowsRuntimeObjectReference? restrictedErrorObject, out bool _))
+            {
+                if (restrictedErrorObject != null)
+                {
+                    IRestrictedErrorInfoMethods.GetErrorDetails(restrictedErrorObject.GetThisPtrUnsafe(), out hr);
+                    GC.KeepAlive(restrictedErrorObject);
```

Comment:
Same here, we need to replace this with an object reference value to make it safe. Also remove `GC.KeepAlive`.

Reply:
I've removed all this code because I realize there is already a duplicate in RestrictedError.cs

https://github.com/microsoft/CsWinRT/pull/2087/commits/081cf64d4a82e8ee905514cea7296b96ff4e7622